providers/AuthProvider.tsx

```diff
@@ -0,0 +1,202 @@
+  login: (email: string, password: string, cb: Function) => void;
+  register: (
+    name: string,
+    email: string,
+    password: string,
+    cb: Function
+  ) => void;
+  loginWithGoogle: (token: string, cb: Function) => void;
+  fetchCurrentUser: () => void;
+};
+
+const AuthContext = createContext<AuthValue | undefined>(undefined);
+
+export const AuthProvider = ({ children }: { children: ReactNode }) => {
+  const fetcher = (url: RequestInfo) => getUserSWR();
+
+  const { data: userSWR, mutate, error } = useSWR("user", fetcher);
+  const [user, setUser] = useState<User | null | undefined>(null);
+  const [status, setStatus] = useState<AuthStatus>("loading");
+
+  const snackbarsService = useSnackbars();
+
+  const router = useRouter();
+
+  useEffect(() => {
+    setStatus("loading");
+    fetchCurrentUser();
+    setUser(userSWR);
+  }, [userSWR]);
+
+  const fetchCurrentUser = () => {
+    getUser((user: User) => {
+      setUser(user);
+      setStatus("connected");
+    }).catch((err: Error) => {
+      setStatus("not-connected");
+
+      // TODO: there is probably a better way to get NotSignedInRoutes
+      if (
+        router.route !== "/" &&
+        router.route !== "/auth/sign-in" &&
+        router.route !== "/auth/sign-up" &&
+        router.route !== "/auth/reset-password"
+      ) {
+        snackbarsService?.addAlert({
+          message: "An error occured while fetching user.",
+          severity: "error",
+        });
+      }
+    });
+  };
+
+  const logout = async () => {
+    await signOut(() => {
+      setUser(null);
+      setStatus("not-connected");
+      router.push("/");
+    }).catch((err: Error) => {
+      setStatus("not-connected");
+      snackbarsService?.addAlert({
+        message: "An error occured while signing out",
+        severity: "error",
+      });
+    });
+  };
+
+  const login = async (email: string, password: string, cb: Function) => {
+    setStatus("loading");
+
+    await signInWithEmailAndPassword(email, password, (user: User) => {
+      cb();
+      setUser(user);
+      setStatus("connected");
+      snackbarsService?.addAlert({
+        message: !user.lastLogin ? "Welcome to ImLazy app !" : "Welcome back !",
+        severity: "success",
+      });
+
+      router.push("/dashboard");
+    }).catch((err: Error) => {
+      cb();
+
+      setUser(null);
+      setStatus("not-connected");
+      //TODO: add internet connection checker and customize message error
+      snackbarsService?.addAlert({
+        message:
+          "Check your internet connection or email/password might be invalid",
+        severity: "error",
+      });
+    });
+  };
+
+  const loginWithGoogle = async (token: string, cb: Function) => {
+    setStatus("loading");
+    await signInWithGoogle(token, (user: User) => {
+      cb();
+      setUser(user);
+      setStatus("connected");
+
+      snackbarsService?.addAlert({
+        message: !user.lastLogin ? "Welcome to ImLazy app !" : "Welcome back !",
```

Comment:
Fais des tests explicites
```suggestion
        message: user.lastLogin === undefined ? "Welcome to ImLazy app !" : "Welcome back !",
```